File: Finales/Septiembre2020/haskell/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin capacity = B capacity []

remainingCapacity :: Bin -> Capacity
remainingCapacity (B cap xs) = cap

addObject :: Weight -> Bin -> Bin
addObject w (B cap xs)
            | w > cap   = error "El objeto no cabe en el cubo"
            | otherwise = (B (cap-w) (w:xs))

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = error "Empty tree."
maxRemainingCapacity (Node bin h cap izq der) = cap

height :: AVL -> Int
height Empty = error "Empty tree."
height (Node bin h cap izq der) = h


 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight (B cap xs) h left right = Node (B cap xs) h (maximum [cap - sum xs, maxRemainingCapacity left, maxRemainingCapacity right]) left right


node :: Bin -> AVL -> AVL -> AVL
node (B cap xs) izq der = (Node (B cap xs) (1 + max (height izq) (height der)) (maximum [cap - sum xs, maxRemainingCapacity izq, maxRemainingCapacity der]) izq der)

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft c ele (Node b h cap rl rr) = node b rl (node c ele rr)

addNewBin :: Bin -> AVL -> AVL
addNewBin b Empty = Node b 1 (remainingCapacity b) Empty Empty
addNewBin b (Node bin h cap left right)
    | height left > height right + 1 = rotateLeft bin left (addNewBin b right)
    | otherwise = node bin left (addNewBin b right)
 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst _ _ Empty = Empty
addFirst _ w (Node (B cap xs) h maxCap left right)
    | w > cap = addNewBin (emptyBin w) (Node (B cap xs) h maxCap left right)
    | maxCap >= w = node (addObject w (B cap xs)) left right
    | maxRemainingCapacity left >= w = node (B cap xs) (addFirst cap w left) right
    | otherwise = node (B cap xs) left (addFirst cap w right)

addAll:: Capacity -> [Weight] -> AVL
addAll _ [] = Empty
addAll w (x:xs) = addFirst w x (addAll w xs)

toList :: AVL -> [Bin]
toList Empty = []
toList (Node bin _ _ left right) = toList left ++ [bin] ++ toList right

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: EstudioVerano/Exámenes/2020 Septiembre/java/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.List;
import dataStructures.list.LinkedList;
import dataStructures.list.ListException;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        weights = new LinkedList<>();
        remainingCapacity = initialCapacity;
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        int sum = 0;

        for(Integer i : weights){
            sum+=i;
        }

        return remainingCapacity - sum;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        if(weight > remainingCapacity()){
            throw new ListException("Error, el objeto no entra en el cubo.");
        }

        weights.append(weight);
        remainingCapacity-=weight;

    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            height = 1 + Math.max(left != null ? left.height : 0, right != null ? right.height : 0);
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {

            maxRemainingCapacity = Math.max(bin.remainingCapacity(), Math.max(left != null ? left.maxRemainingCapacity : 0, right != null ? right.maxRemainingCapacity : 0));

        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {

            Node nuevo = this.right;
            this.right = nuevo.left;
            nuevo.left = this;

            this.setHeight();
            this.setMaxRemainingCapacity();
            nuevo.setMaxRemainingCapacity();
            nuevo.setHeight();

            return nuevo;
        }
    }

    private static int height(Node node) {

        return node.height;
    }

    private static int maxRemainingCapacity(Node node) {
        int max = node.bin.remainingCapacity();

        int leftMax = node.left != null ? node.left.maxRemainingCapacity : 0;
        int rightMax = node.right != null ? node.right.maxRemainingCapacity : 0;

        return Math.max(max, Math.max(leftMax, rightMax));
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        // todo
    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        // todo
    }

    public void addAll(int initialCapacity, int[] weights) {
        // todo
    }

    public List<Bin> toList() {
        // todo
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }
	
	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;		
	}
}